README.md
# AndroidUtils

Some helper code for Android app development.

* [`AndroidUtils`](androidutils/src/main/java/com/uwetrottmann/androidutils/AndroidUtils.kt)
  helper methods to
  * check for Android version (like `AndroidUtils.isLollipopOrHigher()`),
  * an active or unmetered network connection,
  * external storage state availability.
* [`GenericSimpleLoader`](androidutils/src/main/java/com/uwetrottmann/androidutils/GenericSimpleLoader.kt)
  offers a basic `AsyncTaskLoader` implementation to load simple objects or a list of objects.
  * Note: new code should probably use [ViewModel and LiveData](https://developer.android.com/topic/libraries/architecture/livedata) 
    or [Kotlin Flows](https://developer.android.com/kotlin/flow) instead.

## Usage
This library is available on Maven Central.

<a href="https://search.maven.org/search?q=com.uwetrottmann.androidutils"><img src="https://img.shields.io/maven-central/v/com.uwetrottmann.androidutils/androidutils.svg?style=flat-square"></a>

```groovy
implementation("com.uwetrottmann.androidutils:androidutils:4.0.0")
```

## License

This work by [Uwe Trottmann](https://uwetrottmann.com) is licensed under the [Apache License 2.0](LICENSE.txt).

[Contributors](https://github.com/UweTrottmann/AndroidUtils/graphs/contributors) and changes are tracked by Git.

Do not just copy, make it better.


RELEASE_NOTES.md
Release Notes
=============

## 4.0.0
_2023-08-16_

- Add `AndroidUtils.isAtLeastUpsideDownCake`.
- Remove `CheatSheet` (use `androidx.appcompat.widget.TooltipCompat` as a replacement).
- Update Kotlin to 1.8.22. When using this library need to compile with at least Kotlin 1.7.

## 3.1.0
_2022-12-09_

- Fix `AndroidUtils.isNetworkConnected` and `.isUnmeteredNetworkConnected` crashing with 
  `SecurityException` on some Android 11 devices due to a [system bug](https://android-review.googlesource.com/c/platform/frameworks/base/+/1758029).
  See method for details.
- Update Kotlin to 1.7.21, when using the library need to compile with at least Kotlin 1.6.
- Add `AndroidUtils.isAtLeastTiramisu`, `AndroidUtils.isAtLeastSv2`.

## 3.0.0
_2022-01-21_

- Convert to Kotlin.
- Add replace with code for `CheatSheet`, deprecate all methods.
- Add `AndroidUtils.isAtLeastS`, remove `isICSMR1OrHigher`.
- `AndroidUtils` uses `NetworkCapabilities` API to detect network state on Android 6 and higher.
- Remove `AndroidUtils.copyFile`, use a Kotlin extension function instead.

## 2.4.1
_2020-09-18_

- Change new Android SDK version check names to include code-name to appease Lint NewApi check.

## 2.4.0
_2020-09-17_

- Migrate to AndroidX.
- Change `CheatSheet` to internally use `TooltipCompat` from AndroidX appcompat.
- Add Android API level checks up to Android 11 (API level 30).
- Add nullability, permission and type annotations.

## 2.3.1
_2017-08-03_

- `AndroidUtils.copyFile` ensures both streams are closed, no longer throws if closing streams fails. Thanks @NightlyNexus!
- Depend on Android Support library 26.0.0.

## 2.3.0
_2017-05-24_

- Update to `support-core-utils:25.3.1`. No longer depend on whole `support-v4`.
- `minSdk=15` and `targetSdk=25`, added version helper methods for N and N MR1.

## 2.2.1
_2016-05-26_

- Undeprecate version helpers.
- Update support-v4 dependency to 23.4.0.

## 2.2.0
_2016-03-11_

- Removed `executeOnPool()`, use `android.support.v4.os.AsyncTaskCompat` instead.
- Deprecated version helpers, use Studio quick fix to insert `Build.VERSION.SDK_INT` condition instead.

## 2.1.0
_2015-09-09_

- Add version detectors for Android M, LOLLIPOP_MR1, KITKAT_WATCH, JELLY_BEAN_MR2, ICE_CREAM_SANDWICH_MR1.
- Drop FROYO version detector, is above minimal support of this library.
- `isWifiConnected()` now checks if WiFi is actually the active network connection (default route).
- Added `isUnmeteredNetworkConnected()` to check if there is an active and connected network that is not metered, suitable for large data transfer.

## 2.0.0
_2015-05-04_

- Initial deployment to Maven Central.

androidutils/src/main/java/com/uwetrottmann/androidutils/AndroidUtils.java
package com.uwetrottmann.androidutils;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Environment;
import android.text.TextUtils;
import android.view.View;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresPermission;

/**
 * Provides various static helper methods to
 * <ul>
 * <li>check for Android version (like {@link #isAtLeastAndroid8()}),</li>
 * <li>an active or unmetered network connection,</li>
 * <li>external storage state availability.</li>
 * </ul>
 */
public class AndroidUtils {

    /**
     * API level 30+.
     */
    public static boolean isAtLeastAndroid11() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R;
    }

    /**
     * API level 29+.
     */
    public static boolean isAtLeastAndroid10() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q;
    }

    /**
     * API level 28+.
     */
    public static boolean isAtLeastAndroid9() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.P;
    }

    /**
     * API level 27+.
     */
    public static boolean isAtLeastAndroid81() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1;
    }

    /**
     * API level 26+.
     */
    public static boolean isAtLeastAndroid8() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;
    }

    /**
     * API level 25+.
     */
    public static boolean isNougatMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1;
    }

    /**
     * API level 24+.
     */
    public static boolean isNougatOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N;
    }

    /**
     * API level 23+.
     */
    public static boolean isMarshmallowOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
    }

    /**
     * API level 22+.
     */
    public static boolean isLollipopMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1;
    }

    /**
     * API level 21+.
     */
    public static boolean isLollipopOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

    /**
     * API level 20+.
     */
    public static boolean isKitKatWatchOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH;
    }

    /**
     * API level 19+.
     */
    public static boolean isKitKatOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
    }

    /**
     * API level 18+.
     */
    public static boolean isJellyBeanMR2OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2;
    }

    /**
     * API level 17+.
     */
    public static boolean isJellyBeanMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1;
    }

    /**
     * API level 16+.
     */
    public static boolean isJellyBeanOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN;
    }

    /**
     * API level 15+.
     */
    public static boolean isICSMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1;
    }

    /**
     * Checks if {@link Environment}.MEDIA_MOUNTED is returned by {@code getExternalStorageState()}
     * and therefore external storage is read- and writeable.
     */
    public static boolean isExtStorageAvailable() {
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    /**
     * Returns true if the current layout direction is {@link View#LAYOUT_DIRECTION_RTL}.
     *
     * <p>This always returns false on versions below JELLY_BEAN_MR1.
     */
    public static boolean isRtlLayout() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            int direction = TextUtils.getLayoutDirectionFromLocale(Locale.getDefault());
            return direction == View.LAYOUT_DIRECTION_RTL;
        }
        return false;
    }

    @Nullable
    private static ConnectivityManager getConnectivityManager(@NonNull Context context) {
        // use application context to avoid leaking any activity
        // https://code.google.com/p/android/issues/detail?id=198852
        return (ConnectivityManager) context.getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @Nullable
    private static NetworkInfo getActiveNetworkInfo(@NonNull Context context) {
        ConnectivityManager connectivityManager = getConnectivityManager(context);
        if (connectivityManager == null) {
            return null;
        }
        return connectivityManager.getActiveNetworkInfo();
    }

    /**
     * Whether there is an active network connection.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    public static boolean isNetworkConnected(@NonNull Context context) {
        NetworkInfo activeNetworkInfo = getActiveNetworkInfo(context);
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    /**
     * Whether there is an active network connection and it is via WiFi.
     *
     * <p>If you want to check whether to transmit large amounts of data, you may want to use {@link
     * #isUnmeteredNetworkConnected(Context)}.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    public static boolean isWifiConnected(@NonNull Context context) {
        NetworkInfo activeNetwork = getActiveNetworkInfo(context);
        return activeNetwork != null && activeNetwork.isConnected()
                && activeNetwork.getType() == ConnectivityManager.TYPE_WIFI;
    }

    /**
     * Whether there is an active network connection and it is not metered, e.g. so large amounts of
     * data may be transmitted.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public static boolean isUnmeteredNetworkConnected(@NonNull Context context) {
        ConnectivityManager connectivityManager = getConnectivityManager(context);
        if (connectivityManager == null) {
            return false;
        }
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected()
                && !connectivityManager.isActiveNetworkMetered();
    }

    /**
     * Copies the contents of one file to the other using {@link FileChannel}s.
     *
     * @param src source {@link File}
     * @param dst destination {@link File}
     */
    public static void copyFile(@NonNull File src, @NonNull  File dst) throws IOException {
        FileInputStream in = new FileInputStream(src);
        FileOutputStream out = new FileOutputStream(dst);
        FileChannel inChannel = in.getChannel();
        FileChannel outChannel = out.getChannel();
        try {
            inChannel.transferTo(0, inChannel.size(), outChannel);
        } finally {
            try {
                in.close();
            } catch (IOException ignored) {
            }
            try {
                out.close();
            } catch (IOException ignored) {
            }
        }
    }
}
